src/azure-cli/azure/cli/command_modules/network/custom.py

```diff
@@ -434,6 +456,275 @@ def show_ag_identity(cmd, resource_group_name, application_gateway_name):
     return ag.identity
 
 
+def add_ag_private_link(cmd,
+                        resource_group_name,
+                        application_gateway_name,
+                        frontend_ip,
+                        private_link_name,
+                        private_link_subnet_name,
+                        private_link_subnet_prefix,
+                        private_link_primary=None,
+                        private_link_ip_address=None):
+    (SubResource, IPAllocationMethod, Subnet,
+     ApplicationGatewayPrivateLinkConfiguration,
+     ApplicationGatewayPrivateLinkIpConfiguration) = cmd.get_models(
+         'SubResource', 'IPAllocationMethod', 'Subnet',
+         'ApplicationGatewayPrivateLinkConfiguration', 'ApplicationGatewayPrivateLinkIpConfiguration')
+
+    ncf = network_client_factory(cmd.cli_ctx)
+
+    appgw = ncf.application_gateways.get(resource_group_name, application_gateway_name)
+    private_link_config_id = resource_id(
+        subscription=get_subscription_id(cmd.cli_ctx),
+        resource_group=resource_group_name,
+        namespace='Microsoft.Network',
+        type='applicationGateways',
+        name=appgw.name,
+        child_type_1='privateLinkConfigurations',
+        child_name_1=private_link_name
+    )
+
+    if not any(fic for fic in appgw.frontend_ip_configurations if fic.name == frontend_ip):
+        raise CLIError("Frontend IP doesn't exist")
+
+    for fic in appgw.frontend_ip_configurations:
+        if fic.private_link_configuration and fic.private_link_configuration.id == private_link_config_id:
+            raise CLIError('Frontend IP already reference an existing Private Link')
+        if fic.name == frontend_ip:
+            break
+    else:
+        raise CLIError("Frontend IP doesn't exist")
+
+    for pl in appgw.private_link_configurations:
+        if pl.name == private_link_name:
+            raise CLIError('Private Link name duplicates')
+
+    # get the virtual network of this application gateway
+    vnet_name = parse_resource_id(appgw.gateway_ip_configurations[0].subnet.id)['name']
+    vnet = ncf.virtual_networks.get(resource_group_name, vnet_name)
+
+    # prepare the subnet for new private link
+    for subnet in vnet.subnets:
+        if subnet.name == private_link_subnet_name:
+            raise CLIError('Subnet name duplicates')
+        if subnet.address_prefix == private_link_subnet_prefix:
+            raise CLIError('Subnet prefix duplicates')
+        if subnet.address_prefixes and private_link_subnet_prefix in subnet.address_prefixes:
+            raise CLIError('Subnet prefix duplicates')
+
+    private_link_subnet = Subnet(name=private_link_subnet_name,
+                                 address_prefix=private_link_subnet_prefix,
+                                 private_link_service_network_policies='Disabled')
+    vnet.subnets.append(private_link_subnet)
+    ncf.virtual_networks.create_or_update(resource_group_name, vnet_name, vnet)
```

Comment:
I think user should be able to create subnet by himself.

Reply:
Will fix in S173